### PR TITLE
MB-54131: Geoshape query decode optimization

### PR DESCRIPTION
**Pull request by Likith101**
- Added buffer pool pointer to polygon
	- Added parent polygon pointer to loop
	- Added decoder function to select, populate and return a pointer to the buffer from a pool of buffers
	- Updated loop decode function to handle chunks of data being read
	- Updated polygon decode function to pass on the buffer pool to the loops
	- Updated extractShapesFromBytes and FilterGeoShapesOnRelation to include the buffer pool in its parameters

Related PRs
        - https://github.com/blevesearch/bleve/pull/1864

### CODE REVIEW

File: geojson/geojson_shapes_util.go [hunk 1]
```diff
@@ -29,11 +29,12 @@ var jsoniter = jsoniterator.ConfigCompatibleWithStandardLibrary
 
 // FilterGeoShapesOnRelation extracts the shapes in the document, apply
 // the `relation` filter and confirms whether the shape in the document
-//  satisfies the given relation.
+//
+//	satisfies the given relation.
```
Comment (abhinavdangeti) [inline geojson/geojson_shapes_util.go]: Undo this indent, `go fmt ./...` must've changed this because of the unnecessary white space before `satisfies` in line 32 above.

File: geojson/geojson_shapes_util.go [hunk 2]
```diff
@@ -29,11 +29,12 @@ var jsoniter = jsoniterator.ConfigCompatibleWithStandardLibrary
...
 func FilterGeoShapesOnRelation(shape index.GeoJSON, targetShapeBytes []byte,
-	relation string, reader **bytes.Reader) (bool, error) {
+	relation string, reader **bytes.Reader, bufPool *[][]byte) (bool, error) {
```
Comment (abhinavdangeti) [inline geojson/geojson_shapes_util.go]: Does this need to a pointer and not simply [][]byte ?

File: s2/loop.go [hunk 3]
```diff
@@ -413,12 +415,12 @@ func (l *Loop) BoundaryEqual(o *Loop) bool {
...
-//   Then this loop contains XY if and only if Reversed(XY) == the other loop is a hole.
-//   (Intuitively, this checks whether this loop contains a vanishingly small region
-//   extending from the boundary of the other toward the interior of the polygon to
-//   which the other belongs.)
+//	If XY is a shared edge, define Reversed(XY) to be true if XY
+//	  appears in opposite directions in both loops.
+//	Then this loop contains XY if and only if Reversed(XY) == the other loop is a hole.
+//	(Intuitively, this checks whether this loop contains a vanishingly small region
+//	extending from the boundary of the other toward the interior of the polygon to
+//	which the other belongs.)
```
Comment (abhinavdangeti) [inline s2/loop.go]: Same as above, undo this unrelated indentation change.

File: s2/loop.go [hunk 4]
```diff
@@ -979,21 +981,23 @@ func (l *Loop) ContainsNested(other *Loop) bool {
...
 //
-//  - If f is non-negative, then it is easy to detect when the integral over
-//    the loop exterior has been returned, and the integral over the loop
-//    interior can be obtained by adding the integral of f over the entire
-//    unit sphere (a constant) to the result.
+//   - If f is non-negative, then it is easy to detect when the integral over
+//     the loop exterior has been returned, and the integral over the loop
+//     interior can be obtained by adding the integral of f over the entire
+//     unit sphere (a constant) to the result.
 //
```
Comment (abhinavdangeti) [inline s2/loop.go]: All the above as well.

File: s2/loop.go [hunk 5]
```diff
@@ -1287,11 +1291,26 @@ func (l *Loop) decode(d *decoder) {
 		return
 	}
 	l.vertices = make([]Point, nvertices)
-	for i := range l.vertices {
-		l.vertices[i].X = d.readFloat64()
-		l.vertices[i].Y = d.readFloat64()
-		l.vertices[i].Z = d.readFloat64()
+
+	pointsNeeded := int(nvertices) * 3
```
Comment (abhinavdangeti) [inline s2/loop.go]: Add a comment here on why `3`.

File: s2/loop.go [hunk 6]
```diff
@@ -1748,10 +1767,12 @@ func (i *intersectsRelation) wedgesCross(a0, ab1, a2, b0, b2 Point) bool {
...
-//   If A contains a point of both B and ~B, then A intersects Boundary(B).
-//   If ~A contains a point of both B and ~B, then ~A intersects Boundary(B).
-//   So if the intersections of {A, ~A} with {B, ~B} are all non-empty,
-//   the return value is 0, i.e., Boundary(A) intersects Boundary(B).
+//
+//	If A contains a point of both B and ~B, then A intersects Boundary(B).
+//	If ~A contains a point of both B and ~B, then ~A intersects Boundary(B).
+//	So if the intersections of {A, ~A} with {B, ~B} are all non-empty,
+//	the return value is 0, i.e., Boundary(A) intersects Boundary(B).
+//
```
Comment (abhinavdangeti) [inline s2/loop.go]: Undo unrelated changes.

File: s2/polygon.go [hunk 7]
```diff
@@ -38,16 +38,16 @@ import (
...
+//   - Loops may not share edges, i.e. if a loop contains an edge AB, then
+//     no other loop may contain AB or BA.
 //
-//  - Loops may share vertices, however no vertex may appear twice in a
-//    single loop (see Loop).
+//   - Loops may share vertices, however no vertex may appear twice in a
+//     single loop (see Loop).
 //
-//  - No loop may be empty. The full loop may appear only in the full polygon.
+//   - No loop may be empty. The full loop may appear only in the full polygon.
```
Comment (abhinavdangeti) [inline s2/polygon.go]: ditto.

File: s2/polygon.go [hunk 8]
```diff
@@ -77,6 +77,8 @@ type Polygon struct {
 	// preceding loops in the polygon. This field is used for polygons that
 	// have a large number of loops, and may be empty for polygons with few loops.
 	cumulativeEdges []int
+
+	BufPool *[][]byte
```
Comment (abhinavdangeti) [inline s2/polygon.go]: This should perhaps be just `[][]byte`.

**Review by abhinavdangeti — CHANGES_REQUESTED** [review]

File: s2/encode.go [hunk 9]
```diff
@@ -222,3 +222,22 @@ func (d *decoder) readUvarint() (x uint64) {
...
 }
+
+func (d *decoder) readFloat64Array(size int, buffers [][]byte) ([]byte, int) {
+	if d.err != nil {
+		return nil, 0
+	}
+
+	var buf []byte
+
+	for _, buffer := range buffers {
```
Comment (moshaad7) [inline s2/encode.go]: instead of passing around [][]byte as buffers, can we create a new type as GeoBuffPool or something ? 

That way, we can move this logic of selecting a best-fit buffer to some method of that new type.

Thoughts ?

File: s2/encode.go [hunk 10]
```diff
@@ -222,3 +222,22 @@ func (d *decoder) readUvarint() (x uint64) {
...
+		if len(buffer) <= size*8 {
+			buf = buffer
+			break
+		}
+	}
+
+	_, d.err = io.ReadFull(d.r, buf)
+
+	return buf, len(buf) / 8
```
Comment (moshaad7) [inline s2/encode.go]: In a pool design, 
you have to make sure that, while a consumer is using an object from the pool, other consumer shouldn't be able to acquire that very same object ( until the first consume return the object back to the pool).

In our case,
If calls to Decode aren't serialised, then we can end-up in situation where multiple decoders are using same buffer.

File: s2/loop.go [hunk 11]
```diff
@@ -1287,11 +1289,26 @@ func (l *Loop) decode(d *decoder) {
...
+
+	i := 0
+
+	for pointsNeeded > 0 {
+		arr, pointsRead := d.readFloat64Array(pointsNeeded, l.polygon.BufPool)
+		if pointsRead == 0 {
+			break
+		}
+		pointsNeeded = pointsNeeded - pointsRead
+		for j := 0; j < int(pointsRead/3); j++ {
```
Comment (moshaad7) [inline s2/loop.go]: We have to make sure that pointsRead is a multiple of 3, 
( I am assuming that since you are creating buffers with sizes as multiple of 3, formula: 8*x*3 , so won't run into a situation where pointsRead isn't multiple of 3 )

File: s2/buffer_pool.go [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+
+type GeoBufferPool struct {
+	buffers [][]byte
+	maxSize int
+	minSize int
+}
+
+func NewGeoBufferPool(maxSize int, minSize int) *GeoBufferPool {
+
+
```
Comment (abhinavdangeti) [inline s2/buffer_pool.go]: Lose the 2 empty lines here.

File: s2/buffer_pool.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+	return &GeoBufferPool{
+		buffers: make([][]byte, length),
+		maxSize: maxSize,
+		minSize: minSize,
+	}
+}
+
+func (b *GeoBufferPool) Get(size int) ([]byte) {
+
```
Comment (abhinavdangeti) [inline s2/buffer_pool.go]: Drop empty line.

File: s2/buffer_pool.go [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+	bufSize := b.maxSize
+
+	for i := range b.buffers {
+		if bufSize <= size {
+			b.buffers[i] = make([]byte, bufSize)
```
Comment (abhinavdangeti) [inline s2/buffer_pool.go]: It seems we'd be generating a lot of garbage here especially on re-use. Have you considered using a sync.Pool here for the size buffers - so users can return used buffers and we alloc ONLY if the pool has no free buffers of the particular size?

Comment (Likith101) [reply s2/buffer_pool.go]: That is an error on my part. Changing this to create the buffer only if it doesn't exist

File: s2/buffer_pool.go [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+	maxSize = temp / 2
+
+	return &GeoBufferPool{
+		buffers: make([][]byte, length),
+		maxSize: maxSize,
+		minSize: minSize,
+	}
+}
+
+func (b *GeoBufferPool) Get(size int) ([]byte) {
```
Comment (metonymic-smokey) [inline s2/buffer_pool.go:48]: I think a buffer pool also requires a corresponding Put() function which will add back buffers into the pool so that they can be re-used later.

File: .gitignore [hunk 16]
```diff
@@ -0,0 +1 @@
+.DS_Store
```
Comment (metonymic-smokey) [inline .gitignore:1]: Pls remove this file.

File: s2/buffer_pool.go (hunk 15 above)
Comment (moshaad7) [reply s2/buffer_pool.go:48]: This buffer pool doesn't have a notion of Put.
Because this pool expects Get calls to be serialised.

In future versions we are going to write BufferPool which will be shared across multiple queries, that pool will hold multiple instances of various buffer sizes. 
That pool will be concurrent safe and thus will have the notion of Get and Put.

File: .gitignore (hunk 16 above)
Comment (moshaad7) [reply .gitignore:1]: I think there is a need of .gitignore file here because when I pulled the code locally, mac added .DS_Store file as a git change.
Better than manually removing .DS_Store from "git add" , it's better to put it in a ignore file.

File: s2/loop.go [hunk 17]
```diff
@@ -1287,11 +1291,31 @@ func (l *Loop) decode(d *decoder) {
 		return
 	}
 	l.vertices = make([]Point, nvertices)
-	for i := range l.vertices {
-		l.vertices[i].X = d.readFloat64()
-		l.vertices[i].Y = d.readFloat64()
-		l.vertices[i].Z = d.readFloat64()
+
+	// Each vertex requires 24 bytes of storage
+	numBytesNeeded := int(nvertices) * 24
```
Comment (abhinavdangeti) [inline s2/loop.go]: Let's declare a `const vertexSize = 24` to use here.

File: s2/loop.go [hunk 18]
```diff
@@ -1287,11 +1291,31 @@ func (l *Loop) decode(d *decoder) {
...
+		arr := l.BufPool.Get(numBytesNeeded)
+		arr, numBytesRead := d.readFloat64Array(numBytesNeeded, arr)
+
+		if numBytesRead == 0 {
+			break
+		}
+
+		numBytesNeeded = numBytesNeeded - numBytesRead 
+
+		for j := 0; j < int(numBytesRead/24); j++ {
```
Comment (abhinavdangeti) [inline s2/loop.go]: Replace 24 with `vertexSize`.

File: s2/loop.go [hunk 19]
```diff
@@ -1287,11 +1291,31 @@ func (l *Loop) decode(d *decoder) {
...
+			l.vertices[i+j].X = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3) : 8*(j*3+1)]))
+			l.vertices[i+j].Y = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3+1) : 8*(j*3+2)]))
+			l.vertices[i+j].Z = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3+2) : 8*(j*3+3)]))
+		}
+
+		i = i + int(numBytesRead/24)
```
Comment (abhinavdangeti) [inline s2/loop.go]: numBytesRead/vertexSize

File: s2/loop.go [hunk 20]
```diff
@@ -1287,11 +1291,31 @@ func (l *Loop) decode(d *decoder) {
...
+		if numBytesRead == 0 {
+			break
+		}
+
+		numBytesNeeded = numBytesNeeded - numBytesRead 
+
+		for j := 0; j < int(numBytesRead/24); j++ {
+			l.vertices[i+j].X = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3) : 8*(j*3+1)]))
+			l.vertices[i+j].Y = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3+1) : 8*(j*3+2)]))
+			l.vertices[i+j].Z = math.Float64frombits(binary.LittleEndian.Uint64(arr[8*(j*3+2) : 8*(j*3+3)]))
```
Comment (abhinavdangeti) [inline s2/loop.go]: We'll either need to use consts here for 8 and 3 (or derive them from other constants) and more importantly some commentary around why these numbers.

File: s2/buffer_pool.go [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+	bufSize := b.maxSize
+
+	for i := range b.buffers {
+		if bufSize <= size {
```
Comment (Thejas-bhat) [inline s2/buffer_pool.go]: this is my first parse of the PR, but if the caller asks for a buffer size let's say 2047 (Get(2047)), then would you end up returning a buffer of size 1024 instead of 2048?  like isn't the condition for the returned buffer to be something that's closest to the size input but greater than the same? (so that we have allocated enough space and there won't be out of bounds access etc). please let me know if this makes sense

Comment (Likith101) [reply s2/buffer_pool.go]: The condition is the largest buffer that is less than or equal to the buffer size. So 2047 should return 1024.

Comment (moshaad7) [reply s2/buffer_pool.go]: @Likith101 I kind of agree with what @Thejas-bhat  said.
For a Get(100) I am getting a []byte of size 96. 
I will have to use 2 read system calls to read 100 bytes content from a file ( If I am using a buffer of size 96).
In the buffer pool next bigger item is 96*2=196 bytes item and we should return that to user.

So essentially algorithm should be, 
"smallest buffer with size greater than equal to requested size"

But in case we don't have a buffer which can satisfy the above requirement.
Then in that case, we will return the largest available buffer

Thoughts ?

Comment (moshaad7) [reply s2/buffer_pool.go:52]: But in case we don't have a buffer which can satisfy the above requirement.
Then in that case, we will return the largest available buffer

Comment (moshaad7) [reply s2/buffer_pool.go]:
```
// returns smallest available buffer of size >= requestd size.
// If no such buffer is available, returns largest buffer from the pool
func (b *GeoBufferPool) Get(size int) []byte {
	currBuffSize := b.minSize
	for i := 0; i < len(b.buffers); i++ {
		if currBuffSize >= size {
			if b.buffers[i] == nil {
				b.buffers[i] = make([]byte, currBuffSize)
			}
			return b.buffers[i]
		}

		currBuffSize = currBuffSize * 2
	}

	// return largest buffer from the pool
	if b.buffers[len(b.buffers)-1] == nil {
		b.buffers[len(b.buffers)-1] = make([]byte, b.maxSize)
	}

	return b.buffers[len(b.buffers)-1]
}
```

Comment (abhinavdangeti) [reply s2/buffer_pool.go]: Great points by both @Thejas-bhat and @moshaad7 .

I think it makes sense to use a buffer of size **equal to or greater than** what needs to be read up **_until_** the cap (max buffer size in pool) to avoid multiple reads whenever possible.

File: s2/loop.go [hunk 22]
```diff
@@ -1287,11 +1291,36 @@ func (l *Loop) decode(d *decoder) {
 		return
 	}
 	l.vertices = make([]Point, nvertices)
-	for i := range l.vertices {
-		l.vertices[i].X = d.readFloat64()
-		l.vertices[i].Y = d.readFloat64()
-		l.vertices[i].Z = d.readFloat64()
+	floatSize := 8
+	vertexSize := 3 * floatSize
```
Comment (abhinavdangeti) [inline s2/loop.go]: declare global constants for floatSize (which definitely needs a better name) and vertexSize.

File: s2/buffer_pool.go (hunk 21 above)
Comment (Likith101) [reply s2/buffer_pool.go]: I have changed the implementation so that the pool returns the smallest buffer that can completely store the number of bytes needed. As a consequence, the decoder will receive a buffer of higher capacity than necessary so I have limited the decoder to read only the required amount of bytes so that it does not read other metadata stored at the end of the file in case of a loop, or other loops in case of a polygon

File: s2/encode.go [hunk 23]
```diff
@@ -223,12 +223,16 @@ func (d *decoder) readUvarint() (x uint64) {
...
+		return 0
 	}
 
-	_, d.err = io.ReadFull(d.r, buf)
-
-	return buf, len(buf)
+	if size >= len(buf) {
+		_, d.err = io.ReadFull(d.r, buf)
+		return len(buf)
+	} else {
```
Comment (moshaad7) [inline s2/encode.go:234]: since we are using return statement inside "if" block, thus we don't need an else block.

```
if size >= len(buf) {
  _, d.err = io.ReadFull(d.r, buf)
  return len(buf)
}

_, d.err = io.ReadFull(d.r, buf[0:size])
return size

```

File: s2/loop.go [hunk 24]
```diff
@@ -24,6 +25,9 @@ import (
 	"github.com/golang/geo/s1"
 )
 
+const SizeOfFloat = 8
```
Comment (abhinavdangeti) [inline s2/loop.go:28]: Should perhaps use reflect to detect size of float64. Will add separately.

**Review by abhinavdangeti — APPROVED** [review]
